storage/fanout.go

```diff
@@ -696,32 +720,29 @@ func (h *samplesIteratorHeap) Pop() interface{} {
 	return x
 }
 
-// VerticalChunkMergeFunc represents a function that merges multiple time overlapping chunks.
-// Passed chunks:
-// * have to be sorted by MinTime.
-// * have to be part of exactly the same timeseries.
-// * have to be populated.
-type VerticalChunksMergeFunc func(chks ...chunks.Meta) chunks.Iterator
-
 type verticalChunkSeriesMerger struct {
-	verticalChunksMerger VerticalChunksMergeFunc
-
-	labels labels.Labels
-	series []ChunkSeries
+	mergeFunc VerticalSeriesMergeFunc
+	labels    labels.Labels
+	series    []ChunkSeries
 }
 
-// NewVerticalChunkSeriesMerger returns VerticalChunkSeriesMerger that merges the same chunk series into one or more chunks.
-// In case of the chunk overlap, given VerticalChunkMergeFunc will be used.
+// NewCompactingChunkSeriesMerger returns VerticalChunkSeriesMergeFunc that merges the same chunk series into one chunk series.
+// In case of the chunk overlaps, it compacts overlapping chunks into one or more time-ordered chunks with merged data.
+// Samples from overlapped chunks are merged using *series* (not chunk) vertical merge func.
 // It expects the same labels for each given series.
-func NewVerticalChunkSeriesMerger(chunkMerger VerticalChunksMergeFunc) VerticalChunkSeriesMergerFunc {
+//
+// TODO(bwplotka): Currently we have only ChainingSeriesMerge as overlapSeriesMergeFunc. This works the best with
+// replicated series, where data from two series are exactly the same. This does not work well with "almost" the same
+// data e.g from 2 Prometheus HA replicas. Propose penalty based algorithm from Thanos as alternative.
```

Comment:
From a Prometheus standpoint, non-identical data is not a sane setup for remote read. A merge must be a merge, without losing samples.

Reply:
Yes agree. I just propose to add some alternative that will work for all cases. Will add your mentioned use case to comment.

Reply:
We shouldn't be documenting Thanos design decisions that don't apply to Prometheus inside Prometheus.

Reply:
Generally agree, but what if someone want to merge two blocks from different Prometheus replicas (fair use case) without even Thanos?

Reply:
I can remove Thanos mention but will mention about some potentiall improvement we can make for such use case, WDYT?

Reply:
Then they're going to have unsuable data, that it may be inefficient doesn't really matter as the use case isn't sane.

Reply:
Fair. Added comment to `ChainingSeriesMerge` instead. (: